sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala

```diff
@@ -163,6 +163,155 @@ class CollationStringExpressionsSuite
+    val checks = Seq("utf8_binary_lcase", "utf8_binary", "unicode", "unicode_ci").flatMap(
+      c => Seq(
+        SubstringTestCase("select left(null collate " + c + ", 1)", c, Row(null)),
+        SubstringTestCase("select right(null collate " + c + ", 1)", c, Row(null)),
+        SubstringTestCase("select substr(null collate " + c + ", 1)", c, Row(null)),
+        SubstringTestCase("select substr(null collate " + c + ", 1, 1)", c, Row(null)),
+        SubstringTestCase("select left('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select right('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select substr('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select substr('' collate " + c + ", null, null)", c, Row(null))
+      )
+    )
+    checks.foreach(check => {
+      // Result & data type
+      checkAnswer(sql(check.query), check.result)
+      assert(sql(check.query).schema.fields.head.dataType.sameType(StringType(check.collation)))
+    })
+  }
+
+  test("Support Left/Right/Substr with explicit improper length & position") {
+    case class SubstringTestCase(query: String, collation: String, result: Row)
+    val checks = Seq("utf8_binary_lcase", "utf8_binary", "unicode", "unicode_ci").flatMap(
+      c => Seq(
+        SubstringTestCase("select left(' a世a ' collate " + c + ", '3')", c, Row(" a世")), // scalastyle:ignore
+        SubstringTestCase("select right(' a世a ' collate " + c + ", '3')", c, Row("世a ")), // scalastyle:ignore
+        SubstringTestCase("select right('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select substr('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select substr('' collate " + c + ", null, null)", c, Row(null)),
+        SubstringTestCase("select left('' collate " + c + ", null)", c, Row(null))
+      )
+    )
+    checks.foreach(check => {
+      // Result & data type
+      checkAnswer(sql(check.query), check.result)
+      assert(sql(check.query).schema.fields.head.dataType.sameType(StringType(check.collation)))
+    })
+  }
+
+  test("Support Left/Right/Substr with session collation") {
+    case class SubstringTestCase(query: String, collation: String, result: Row)
+    val checks = Seq("utf8_binary_lcase", "utf8_binary", "unicode", "unicode_ci")
+      .flatMap { c =>
+        Seq(
+          SubstringTestCase("select left('abc', 1)", c, Row("a")),
+          SubstringTestCase("select right('def', 1)", c, Row("f")),
+          SubstringTestCase("select substr('ghi', 1)", c, Row("ghi"))
+        )
+      }
+    checks.foreach(check => {
+      withSQLConf(SqlApiConf.DEFAULT_COLLATION -> check.collation) {
+        // Result & data type
+        checkAnswer(sql(check.query), check.result)
+        assert(sql(check.query).schema.fields.head.dataType.sameType(StringType(check.collation)))
+      }
+    })
+  }
```

Comment:
I think this test is not needed

it's good that you verified that this works, but I think session default collation is a separate effort

we shouldn't need to test whether every single string expression works with DEFAULT_COLLATION, because there's nothing specific about them

Reply:
Will remove!